tsdb/index.go

```diff
@@ -1699,19 +1822,25 @@ func (is IndexSet) measurementNamesByTagFilter(auth query.FineAuthorizer, op inf
 	}
 
 	bytesutil.Sort(names)
-	return names, nil
+	failed = false
+	return newFileMeasurementSliceIterator(names, MeasurementIterators{mitr}), nil
 }
 
-func (is IndexSet) measurementNamesByTagPredicate(auth query.FineAuthorizer, op influxql.Token, key, val string, regex *regexp.Regexp) ([][]byte, error) {
+func (is IndexSet) measurementNamesByTagPredicate(auth query.FineAuthorizer, op influxql.Token, key, val string, regex *regexp.Regexp) (MeasurementIterator, error) {
 	var names [][]byte
+	failed := true
 
 	mitr, err := is.measurementIterator()
 	if err != nil {
 		return nil, err
 	} else if mitr == nil {
 		return nil, nil
 	}
-	defer mitr.Close()
+	defer func() {
```

Comment:
I've recently seen a pattern using `sync.Once` to do this sort of thing, and I like it.

Reply:
E.g.:

```
// early:
once sync.Once{}
defer once.Do(mitr.Close)

//later
var toReturn ...
once.Do(func(){
  toReturn = newFileMeasurementSliceIterator(names, MeasurementIterators{mitr}), nil
})
return toReturn
```

So you're guaranteed just one of either the close or the delegation of the close succeeds.

Reply:
This comment doesn't need to be taken for the PR, just an observation.

Reply:
I think that this pattern won't work here, because the initial `defer once.Do(mitr.Close)` will be called first, before the close on the `toReturn`. And that's what is causing the panic; a close on the iterator before the data is read out.  But it's a nice pattern for other places.

Reply:
Hmm - I think it would work, here's a playground example. But not necessary here. https://play.golang.org/p/35qO7UPTOyG

Reply:
Ah, I see; Once works differently than a similar construct I used in another language; my mistake.